### PR TITLE
Consistently quote in grammar (and remove one spurious ;).

### LINKED ISSUE #2988: Syntax of preferredView annotation
The specification of the `preferredView` annotation looks inconsistent.

According to the grammar of the annotation, it should be used like this (supported by System Modeler):
```
annotation(preferredView = info);
```

However, in the examples it is used with a string instead, in agreement with how the annotation is used in the MSL:
```
annotation(preferredView = "info");
```

I see some alternatives:
- Pretend the non-string variant has ever existed; just silently change it to use strings instead.
- Add the string form and treat both forms equal.
- Add the string form and deprecate the non-string form.

### PR DESCRIPTION
**Pull request by HansOlsson**
I reviewed all fragments marked as "grammar" based on #2988; and as expected found a few additional problems.

Note that some code fragments are in Modelica and some are in the grammar-language; we might at some point consider to have them more visible different to avoid confusion.
Closes #2988

### CODE REVIEW

File: chapters/annotations.tex
Comment (henrikt-ma) [inline chapters/annotations.tex:1122]: As indicated by the comments in #2988, this could actually be considered a backwards incompatible change.  Therefore, I think we need a language group decision before just throwing out the non-string variant.

Also, I'm not convinced that triple double quotes is the proper way to quote a string literal in the grammar.

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
Most of this looks fine, but I think at least one of the changes might need both tweaking and a language group decision to support it.

File: chapters/annotations.tex
Comment (HansOlsson) [reply chapters/annotations.tex]: I modified it to `""" "info"`  etc. 
It's clear that triple double quotes is the way used in the grammar to generate a double-quote, it's used both in the grammar and already in this section.
And` "info"` is similar to how strings are handled elsewhere for a literal string `info`. 
If we strictly followed the production of string it would be `""" "i" "n" "f" ...` - but that is unreadable.

Comment (HansOlsson) [reply chapters/annotations.tex]: And of course the decision on #2988 is still needed.

Comment (henrikt-ma) [reply chapters/annotations.tex]: > It's clear that triple double quotes is the way used in the grammar to generate a double-quote, it's used both in the grammar and already in this section.

Yes, so far so good; we use double quotes to quote tokens that aren't grammar keywords.  For example:
```
"experiment"
   [ "(" [ experimentOption { "," experimentOption } ] ")" ]
```

The non-quoted names such as _experimentOption_ are production rules and will be typeset as such.

> And` "info"` is similar to how strings are handled elsewhere for a literal string `info`.

If there are examples of that, I'd consider these incorrect.

> If we strictly followed the production of string it would be `""" "i" "n" "f" ...` - but that is unreadable.

It would not only be unreasonable, it could also mean that you could have whitespace inside, which is not the intention:
```
preferredView = " i      n       f      o         "
```

Well, we actually seem to use the grammar language also for lexing definitions where white-space needs to be expressed explicitly.  For example, see 18.8.1 _Version Numbering_.  It seems like the use of upper case names in the production rules could be a hint about expressing lexing units rather than production rules, so maybe we could have done it in this extremely ugly fashion:
```
"preferredView" "=" ( INFO | … )

INFO = """ "i" "n" "f" "o" """
```

By the way, I note that the syntax highlighter for the grammar language doesn't understand that `"""` isn't an empty string followed by the opening of another string.  Fixing this is related to the current problem of expressing string literals.

Anyway, if we really insist on strings and use of the grammar language for expressing the annotation, we should have a way of properly up-quoting a string literal, which is something we might not have done before.  (The less flexible alternative of not using strings simply seems like better fit with our grammar language.)

For example – not saying we should do this, as it gets pretty messy to read – one could use single quotes around the double quotes to indicate a string literal:
- `"x"` – a single character token: not a string literal
- ''"x"' – the one character string literal `"x"`
- ''"info"' – the string literal `"info"`

As a better alternative, we could try avoiding the problem altogether by not using the grammar language at all.  This makes good sense if we don't care about non-string settings such as `preferredView = info`.  We've only done this in the context of pseudo-code records before, but maybe we could also do it without a record:
```
String preferredView;
```
The text should then state that the value must be given by a literal, and what the valid strings are.

**Comment by HansOlsson** [comment]
> The main problem I see is that we also need to deprecate the non-`String` variant when doing this.

That variant was never intended.
It wasn't used in MSL, and doesn't really make sense internally since preferredView is neither quoted nor a keyword.
Additionally the text below for DocumentationClass already in 3.4 said:

> Only allowed as class annotation on any kind of class and implies that this class and all classes within it are treated as having the annotation preferredView="info".

Basically the specification was already a bit messed up, and the conversion to a more formal language=grammar in the LaTeX-conversion didn't help. I don't see that adding some deprecated text is necessary.

File: chapters/unitexpressions.tex
Comment (henrikt-ma) [reply chapters/unitexpressions.tex]: Looks good.  Resolving.

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
> Basically the specification was already a bit messed up, and the conversion to a more formal language=grammar in the LaTeX-conversion didn't help. I don't see that adding some deprecated text is necessary.

Nevertheless, at least System Modeler also allows the "messed up" non-string variant.  So for us, suddenly invalidating all such annotations is a too hard backward incompatible change.

**Comment by HansOlsson** [comment]
> > Basically the specification was already a bit messed up, and the conversion to a more formal language=grammar in the LaTeX-conversion didn't help. I don't see that adding some deprecated text is necessary.
> 
> Nevertheless, at least System Modeler also allows the "messed up" non-string variant. So for us, suddenly invalidating all such annotations is a too hard backward incompatible change.

I don't see anything preventing you from handling such cases in a backwards compatible manner, even if not formally deprecated. We could add "Other values, including non-Strings, are not standardized, but may be handled by tools."

My point is that I don't see how the previous text supports `preferredView=info`. The 3.4 specification contains:

> `annotation"(" preferredView "=" ("info" | "diagram" | "text") ")"`
> ...
> Only allowed as class annotation on any kind of class and implies that this class and all classes within it are 
> treated as having the annotation preferredView="info".

To me this is just a mess, if `"info"` is seen as the text `info` (without quotes) as in the grammar, then for consistency reasons preferredView should be a reserved word, like annotation, which it isn't, and it would be inconsistent with the other text.

**Comment by henrikt-ma** [comment]
> To me this is just a mess, if `"info"` is seen as the text `info` (without quotes) as in the grammar, then for consistency reasons preferredView should be a reserved word, like annotation, which it isn't, and it would be inconsistent with the other text.

It is indeed a mess, and in this mess System Modeler has opted on the side of not rejecting annotations that could be legal according to some more or less crazy interpretation of this mess.  This has resulted in libraries where annotations use the non-string form, and making them illegal without prior deprecation just doesn't seem right.

I don't like "Other values, including non-Strings, are not standardized, but may be handled by tools." since it doesn't give a valid excuse for the existence of the non-strings in libraries that were not developed with the intention of only working in a specific tool, and it also seems like a counter-productive invitation to keep using the "wrong" variant.

**Comment by HansOlsson** [comment]
> I don't like "Other values, including non-Strings, are not standardized, but may be handled by tools." since it doesn't give a valid excuse for the existence of the non-strings in libraries that were not developed with the intention of only working in a specific tool, and it also seems like a counter-productive invitation to keep using the "wrong" variant.

Ok, I added some text for this. But I don't want to add it too much, since that would imply that other tools also need to support it, which wasn't the intention.

File: chapters/annotations.tex
Comment (henrikt-ma) [inline chapters/annotations.tex:1125]: Maybe we should write out "information"?
```suggestion
The $layer$ is a \lstinline!String! literal where \lstinline!"info"! means information layer, i.e., the documentation of the class, \lstinline!"diagram"! means diagram layer and \lstinline!"text"! means the Modelica text layer.
```

File: chapters/annotations.tex
Comment (henrikt-ma) [inline chapters/annotations.tex:1129]: Deprecated means normatively deprecated.  We shouldn't introduce a new habit of having non-normatively deprecated features.  By the way, what would being non-normatively deprecated even mean – more or less a standard language feature compared to other deprecated features?

```suggestion
A deprecated variant of the annotation is to use identifiers instead of \lstinline!String! literals, for example, \lstinline!preferredView = info!.
```

**Review by henrikt-ma — CHANGES_REQUESTED** [review]

File: chapters/annotations.tex
Comment (HansOlsson) [reply chapters/annotations.tex:1129]: Normatively making it a deprecated feature would mean that all tools must support it, until it is removed.

Requiring all tools to support `preferredView=info` does not make sense.

**Comment by henrikt-ma** [comment]
> Ok, I added some text for this. But I don't want to add it too much, since that would imply that other tools also need to support it, which wasn't the intention.

I'm afraid making it a deprecated feature is our best path out of the mess.  I wouldn't expect tools to be overly ambitious to support the deprecated form, but the blame needs to be on the specification, not the users and tools who were given non-intended impressions by a historical mess in the specification.

**Comment by HansOlsson** [comment]
> > Ok, I added some text for this. But I don't want to add it too much, since that would imply that other tools also need to support it, which wasn't the intention.
> 
> I'm afraid making it a deprecated feature is our best path out of the mess.

No, it is not.

The proposed variant state that tools may still support `preferredView=info`, but users should change to `preferredView="info"`.

The other variant requires that _all_ tools must support this for the time being, but users should still upgrade to `preferredView="info"`. That didn't work before in all tools, as not all tools had implemented this.

So, the difference is that tools must add support for a feature that shouldn't be used; that's not time well spent.

We could say that the optional deprecated feature require a diagnostic if that feels better.

> I wouldn't expect tools to be overly ambitious to support the deprecated form, but the blame needs to be on the specification, not the users and tools who were given non-intended impressions by a historical mess in the specification.

It's not about who to blame, but about having a reasonable specification for the future to encourage tools to implement the specification.

**Comment by henrikt-ma** [comment]
> The other variant requires that _all_ tools must support this for the time being, but users should still upgrade to `preferredView="info"`. That didn't work before in all tools, as not all tools had implemented this.

Unfortunate, but hard to avoid within good practices of specification development.  Note that to must support a deprecated part of the specification is not the same thing as going to do it.

It seems like we need a web meeting decision on which formulation to use.

**Comment by HansOlsson** [comment]
> > The other variant requires that _all_ tools must support this for the time being, but users should still upgrade to `preferredView="info"`. That didn't work before in all tools, as not all tools had implemented this.
> 
> Unfortunate, but hard to avoid within good practices of specification development. 

I don't see that it is **good practice** for specifications that mistakes leading to odd interpretations in some tools should be included as required, but deprecated in the specification itself. Do you have any reference for that statement?

**Comment by henrikt-ma** [comment]
> I don't see that it is **good practice** for specifications that mistakes leading to odd interpretations in some tools should be included as required, but deprecated in the specification itself. Do you have any reference for that statement?

No.  I thought it was obvious. :)

**Comment by HansOlsson** [comment]
The remaining issue seems to be the handling of preferredView=info.
I see three possibilities:

1. Don't say anything about info without quotes.
2. Say that tools may as deprecated feature support it
3. Make it mandatory for all tools as a deprecated feature.
I would prefer 1, but can if necessary live with 2.

**Comment by HansOlsson** [comment]
Clear that allowed: annotation(preferredView="info");

What about:
annotation(preferredView=info);
1. Don't say anything about info without quotes. Hans, Martin, Gerd, (Henrik), Markus, Elena, Christoff
2. Say that tools may as deprecated feature support it. Hans, Elena, Gerd
3. Make it a deprecated feature in the language. Henrik
Somewhat agreement.

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
I think we should wait until #3067 is merged before looking this.

**Comment by HansOlsson** [comment]
> I think we should wait until #3067 is merged before looking this.

It's the other way round, I based the #3067 branch on this branch.

File: chapters/annotations.tex
Comment (henrikt-ma) [inline chapters/annotations.tex:1121]:
```suggestion
annotation(preferredView = $\mathit{layer}$)
```

File: chapters/annotations.tex
Comment (henrikt-ma) [inline chapters/annotations.tex:1125]:
```suggestion
The $\mathit{layer}$ is a \lstinline!String! literal where \lstinline!"info"! means information layer, i.e., the documentation of the class, \lstinline!"diagram"! means diagram layer and \lstinline!"text"! means the Modelica text layer.
```

File: chapters/annotations.tex
Comment (henrikt-ma) [reply chapters/annotations.tex:1129]: Remove.  (Not sure why GitHub doesn't let me make a suggestion.)

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
Needs back-porting of stuff from #3067.

**Comment by henrikt-ma** [comment]
> > I think we should wait until #3067 is merged before looking this.
> 
> It's the other way round, I based the #3067 branch on this branch.

Oh.  I've learnt the hard way that this is never a good idea.

**Pull request merged**
